scripts/dss_v2_apdapter.py

```diff
@@ -0,0 +1,316 @@
+        log.debug(f'{round(total_duration, 4)} Total')
+
+    def upload_file_contents(self, file_contents: Mapping[str, Any], new_name: str, content_type: str) -> Blob:
+        """
+        Perform an upload of a dict (json file content) to a bucket
+        """
+        log.info('Uploading to: %s', new_name)
+        self.dst_bucket.blob(new_name).upload_from_string(data=json.dumps(file_contents, indent=4),
+                                                          content_type=content_type)
+        return self.dst_bucket.blob(new_name).exists()
+
+    def copy_file(self, blob_key: str, new_name: str) -> bool:
+        """
+        Perform a bucket to bucket copy
+        """
+        log.info('Copying to: %s', new_name)
+        src_blob = self.src_bucket.blob(f'blobs/{blob_key}')
+        dst_blob = self.src_bucket.copy_blob(blob=src_blob,
+                                             destination_bucket=self.dst_bucket,
+                                             new_name=new_name)
+        return isinstance(dst_blob, Blob)
+
+    def log_error(self, bundle_uuid: str, file_name: str, msg: str):
+        """
+        Log an error message and save a copy for re-reporting at end of script
+        """
+        self.errors.append((bundle_uuid, file_name, msg))
+        log.error(msg)
+
+    @classmethod
+    def links_json_new_name(cls, bundle_uuid: str, bundle_version: str) -> str:
+        """
+        Return the bucket layout compliant object name for a links.json file
+        """
+        return f'{bundle_uuid}_{bundle_version}.json'
+
+    @classmethod
+    def metadata_file_new_name(cls, entity_type: str, entity_id: str, version: str) -> str:
+        """
+        Return the bucket layout compliant object name for a metadata file
+        """
+        return f'{entity_type}/{entity_id}_{version}.json'
+
+    @classmethod
+    def data_file_new_name(cls, old_name: str, file_uuid: str, file_version: str) -> str:
+        """
+        Return the bucket layout compliant object name for a data file
+        """
+        old_name = old_name.replace('!', '/')  # Undo DCP v1 substitution of '/' with '!'
+        assert not old_name.endswith('/')
+        pos = old_name.rfind('/')
+        if pos == -1:
+            dir_path = ''
+        else:
+            dir_path, old_name = old_name[:pos + 1], old_name[pos + 1:]
+        return f'{dir_path}{file_uuid}_{file_version}_{old_name}'
+
+    @classmethod
+    def ascending_prefixes(cls, prefix: str):
+        """
+        Generate ascending hex prefixes starting with given value.
+
+        >>> list(cls.ascending_prefixes('7'))
+        ['7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f']
+
+        >>> list(cls.ascending_prefixes('aaa'))
+        ['aaa', 'aab', 'aac', 'aad', 'aae', 'aaf', 'ab', 'ac', 'ad', 'ae', 'af', 'b', 'c', 'd', 'e', 'f']
+        """
+        chars = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f']
```

Comment:
Note that this is essentially counting up in base 16. Given that your implementation struck me as a little complicated:

![image](https://user-images.githubusercontent.com/5143256/82091824-ef352400-96ac-11ea-840d-b80265b705a0.png)

Reply:
But it doesn't have the `aaf` to `ab` shortening. How do you know you don't shorten beyond the original prefix length?

Reply:
> How do you know you don't shorten beyond the original prefix length?

I'm not sure I understand this question. The original intent of the "start prefix" argument was to allow the user to resume processing bundles from where they last left off after a crash or ctrl-c. While `aaf` to `ab` shortening isn't necessary for this occur, it does allow for a very slight performance increase by reducing the number of calls to get a list of bundles from the DSS `/bundles/all` endpoint.

Reply:
I see. The two bundle prefix arguments were intended to be mutually exclusive. My mistake was in not specifying this with argparse.